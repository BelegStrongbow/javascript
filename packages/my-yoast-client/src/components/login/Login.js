import PropTypes from "prop-types";
import React from "react";
import { intlShape, injectIntl } from "react-intl";
import { redirectToOAuthUrl } from "../../functions/auth";

// Components.
import LoginForm from "./LoginForm";


/**
 * Login component for logging in to MyYoast.
 */
class Login extends React.Component {
	constructor( props ) {
		super( props );

		this.state = {
			email: "",
			password: "",
			rememberMe: false,
			otp: "",
		};

		this.onRememberCheck = this.onRememberCheck.bind( this );
		this.handleSubmit = this.handleSubmit.bind( this );
		this.onUpdateEmail = this.onUpdateField.bind( this, "email" );
		this.onUpdatePassword = this.onUpdateField.bind( this, "password" );
		this.onUpdateOTP = this.onUpdateField.bind( this, "otp" );
	}

	/**
	 * Event listener, listening to changes of the
	 * 'remember me' checkbox (e.g. whether it is checked or not).
	 * @param {Object} event the input change event.
	 * @returns {void}
	 */
	onRememberCheck( event ) {
		this.setState( {
			rememberMe: event.target.checked,
		} );
	}

	/**
	 * Updates the specified field in the state,
	 * to be used as callback functions in text input fields.
	 *
	 * @param {string} field the field in the state that should be updated.
	 * @param {Object} event the input field change event.
	 * @returns {void}
	 */
	onUpdateField( field, event ) {
		const obj = {};
		obj[ field ] = event.target.value;
		this.setState( obj );
	}

	/**
	 * Handles the login submit by attempting a login.
	 *
	 * @param {Object} event The event.
	 *
	 * @returns {void}
	 */
	handleSubmit( event ) {
		event.preventDefault();
		if ( this.props.loading ) {
			return;
		}
		const data = { email: this.state.email, password: this.state.password, rememberMe: this.state.rememberMe, otp: this.state.otp };
		this.props.attemptLogin( data );
	}

	shouldComponentUpdate() {
		if ( this.props.oauthError && this.props.loggedIn ) {
			redirectToOAuthUrl();
			return false;
		}
		return true;
	}

	/**
	 * Renders the component.
	 *
	 * @returns {ReactElement} The rendered component.
	 */
	render() {
		return (
			<LoginForm
				rememberMe={ this.state.rememberMe }
				email={ this.state.email }
				password={ this.state.password }
				errors={ this.props.error }
				loading={ this.props.loading }
				onUpdateEmail={ this.onUpdateEmail }
				onUpdatePassword={ this.onUpdatePassword }
				onUpdateOTP={ this.onUpdateOTP }
				onRememberCheck={ this.onRememberCheck }
				handleSubmit={ this.handleSubmit }
				requireOTP={ this.props.requireOTP }
			/>
		);
	}
}

Login.propTypes = {
	intl: intlShape.isRequired,
	error: PropTypes.object,
	loading: PropTypes.bool,
	loggedIn: PropTypes.bool,
	oauthError: PropTypes.bool,
	requireOTP: PropTypes.bool.isRequired,
	attemptLogin: PropTypes.func.isRequired,
};

Login.defaultProps = {
	error: null,
	loading: false,
	requireOTP: false,
};

export default injectIntl( Login );
