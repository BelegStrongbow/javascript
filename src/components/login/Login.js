import PropTypes from "prop-types";
import React from "react";
import { intlShape, injectIntl } from "react-intl";
import { redirectToOAuthUrl } from "../../functions/auth";

// Components.
import LoginForm from "./LoginForm";


/**
 * Login component for logging in to MyYoast.
 */
class Login extends React.Component {

	constructor( props ) {
		super( props );

		this.state = {
			email: "",
			password: "",
			rememberMe: false,
		};

		this.onRememberCheck = this.onRememberCheck.bind( this );
		this.handleSubmit = this.handleSubmit.bind( this );
		this.onUpdateEmail = this.onUpdateField.bind( this, "email" );
		this.onUpdatePassword = this.onUpdateField.bind( this, "password" );
	}

	/**
	 * Event listener, listening to changes of the
	 * 'remember me' checkbox (e.g. whether it is checked or not).
	 * @param {Object} event the input change event.
	 * @returns {void}
	 */
	onRememberCheck( event ) {
		this.setState( {
			rememberMe: event.target.checked,
		} );
	}

	/**
	 * Updates the specified field in the state,
	 * to be used as callback functions in text input fields.
	 *
	 * @param {string} field the field in the state that should be updated.
	 * @param {Object} event the input field change event.
	 * @returns {void}
	 */
	onUpdateField( field, event ) {
		let obj = {};
		obj[ field ] = event.target.value;
		this.setState( obj );
	}

	/**
	 * Handles the login submit by attempting a login.
	 *
	 * @param {Object} event The event.
	 *
	 * @returns {void}
	 */
	handleSubmit( event ) {
		event.preventDefault();
		if ( this.props.loading ) {
			return;
		}
		let data = { email: this.state.email, password: this.state.password, rememberMe: this.state.rememberMe };
		this.props.attemptLogin( data );
	}

	shouldComponentUpdate() {
		if ( this.props.oauthError && this.props.loggedIn ) {
			redirectToOAuthUrl();
			return false;
		}
		return true;
	}

	render() {
		return (
			<LoginForm
				rememberMe={ this.state.rememberMe }
				email={ this.state.email }
				password={ this.state.password }
				errors={ this.props.error }
				loading={ this.props.loading }
				onUpdateEmail={ this.onUpdateEmail }
				onUpdatePassword={ this.onUpdatePassword }
				onRememberCheck={ this.onRememberCheck }
				handleSubmit={ this.handleSubmit }
			/>
		);
	}
}

Login.propTypes = {
	intl: intlShape.isRequired,
	error: PropTypes.object,
	loading: PropTypes.bool,
	loggedIn: PropTypes.bool,
	oauthError: PropTypes.bool,
	attemptLogin: PropTypes.func.isRequired,
};

Login.defaultProps = {
	error: null,
	loading: false,
};

export default injectIntl( Login );
