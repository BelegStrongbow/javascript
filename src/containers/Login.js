import { connect } from "react-redux";
import Login from "../components/login/Login";
import { loginRequest } from "../actions/login";
import get from "lodash/get";

export const mapStateToProps = ( state ) => {
	const newprops = Object.assign( {}, state.ui.login );
	if ( newprops.amountOfOTPWarnings === 1 && get( newprops, "error.code" ) === "invalid_google_authenticator_token" ) {
		newprops.error = { error: { code: "otp_required" } };
	}
	return newprops;
};

export const mapDispatchToProps = ( dispatch ) => {
	return {
		/**
		 * Opens the door to the treasures of MyYoast,
		 * if their credentials are correctly filled in.
		 *
		 * @param {Object} params The request parameters.
		 * @param {string} params.email The email address of the user who is trying to log in.
		 * @param {string} params.password The password of the user who is trying to log in.
		 * @param {boolean} params.rememberMe Whether or not the user wants to be reminded, and have their login session extended.
		 *
		 * @returns {void}
		 */
		attemptLogin: ( params ) => {
			dispatch( loginRequest( params ) );
		},
	};
};

const LoginContainer = connect(
	mapStateToProps,
	mapDispatchToProps
)( Login );

export default LoginContainer;
