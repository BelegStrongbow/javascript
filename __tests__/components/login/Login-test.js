import React from 'react';
import { createComponentWithIntl } from "../../../utils";

import Login from "../../../src/components/login/Login";

test( 'The Login component matches the snapshot', () => {
	const component = createComponentWithIntl(
		<Login />,
	);

	let tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );