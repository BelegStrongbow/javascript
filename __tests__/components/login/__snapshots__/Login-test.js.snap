// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Login component matches the snapshot 1`] = `
.c14 {
  height: 48px;
  margin: 0;
  padding: 0 16px;
  border: 0;
  background-color: #64a60a;
  color: #fff;
  box-shadow: inset 0 -4px 0 rgba(0,0,0,0.2);
  -webkit-filter: drop-shadow(0 2px 4px rgb(0,0,0,0.2));
  filter: drop-shadow(0 2px 4px rgb(0,0,0,0.2));
  border-radius: 4px;
  font: 700 14px/24px "Open Sans",sans-serif;
  cursor: pointer;
  vertical-align: top;
  -webkit-transition: background-color 150ms ease-out;
  transition: background-color 150ms ease-out;
}

.c14:hover,
.c14:focus {
  box-shadow: 0 4px 10px 0 rgba(0,0,0,0.2),inset 0 0 0 100px rgba( 0,0,0,0.1 );
}

.c14:active {
  -webkit-transform: translateY( 1px );
  -ms-transform: translateY( 1px );
  transform: translateY( 1px );
  box-shadow: none;
}

.c13 {
  text-shadow: 0px 0px 2px #000;
}

.c5 {
  box-shadow: inset 0 1px 8px 0 rgba(0,0,0,0.3);
  background: #fff;
  border: 0;
  padding: 0 0 0 10px;
  font-size: 1em;
  font-family: "Open Sans",sans-serif;
  height: 3em;
  width: 100%;
}

.c8 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  vertical-align: top;
}

.c9 {
  position: absolute;
  opacity: 0;
}

.c9:checked + .checkmark {
  background-color: #a4286a;
  background-image: url( check.svg );
  background-repeat: no-repeat;
  background-position: center;
  background-size: 75%;
}

.c9:focus + .checkmark {
  box-shadow: inset 0 1px 8px 0 rgba(0,0,0,0.3),0 0 2px 2px #1e8cbe;
  outline: 2px solid transparent;
}

.c10 {
  height: 25px;
  width: 25px;
  cursor: pointer;
  box-shadow: inset 0 1px 8px 0 rgba(0,0,0,0.3);
  background-color: #f7f7f7;
}

.c10:hover {
  background-color: #cecece;
}

.c11 {
  cursor: pointer;
  margin-left: 8px;
}

.c3 {
  display: inline-block;
  font-size: 1em;
  margin: 16px 0 8px;
  font-weight: 700;
}

.c4 {
  background-color: #f7f7f7;
}

.c0 {
  position: relative;
  width: 400px;
  min-height: 400px;
}

.c1 {
  width: 100%;
}

.c2 {
  margin-top: 5px;
}

.c6 {
  margin-top: 10px;
}

.c12 {
  margin: 1em 0;
  width: 100%;
}

.c7 {
  margin-top: 20px;
  width: 100%;
  font-weight: 700;
}

<form
  className="c0"
  onSubmit={[Function]}
>
  <div
    className="c1"
  >
    <label
      className="c2 c3"
      htmlFor="email-address"
    >
      <span>
        Email address
      </span>
    </label>
    <input
      className="c4 c5"
      id="email-address"
      name="email"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="c1"
  >
    <label
      className="c2 c3"
      htmlFor="password"
    >
      <span>
        Password
      </span>
    </label>
    <input
      className="c4 c5"
      id="password"
      name="password"
      onChange={[Function]}
      type="password"
    />
  </div>
  <div
    className="c6"
  >
    <a
      href="/forgot-password"
    >
      <span>
        Forgot your password?
      </span>
    </a>
  </div>
  <div
    className="c7"
  >
    <div
      className="c8"
    >
      <input
        checked={false}
        className="c9"
        id="remember-me"
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="checkmark c10"
      />
      <label
        className="c11"
        htmlFor="remember-me"
      >
        Remember me
      </label>
    </div>
  </div>
  <div>
    <button
      className="c12 c13 c14"
      type="submit"
    >
      <span>
        Log in
      </span>
    </button>
  </div>
</form>
`;
